include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,168 @@ module BYTES-SIMPLIFICATION [symbolic]
+      #range (#buf(W1:Int, X:Int), S2:Int, W2:Int) => #buf(W2, 0)
+      requires 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int (S2 +Int W2))) andBool S2 +Int W2 <=Int W1
+      [simplification]
 
-    // lookup range in memory
-    rule #range(MEM, 0, W) => MEM  requires W ==Int lengthBytes(MEM) [simplification]
+    rule [range-padRightToWidth]:
+      #range(#padRightToWidth(_, BUF), 0, WIDTH) => BUF
+      requires lengthBytes(BUF) ==Int WIDTH [simplification]
 
-    // lookup range in concatenated memory buffers
-    rule #range((BUF1 +Bytes  BUF2), START, WIDTH) =>  #range(BUF2, START -Int lengthBytes(BUF1), WIDTH                        )                                                                    requires                        lengthBytes(BUF1) <=Int START                                                          [simplification]
-    rule #range((BUF1 +Bytes  _   ), START, WIDTH) =>  #range(BUF1, START                       , WIDTH                        )                                                                    requires START +Int WIDTH <=Int lengthBytes(BUF1)                       andBool 0 <=Int START andBool 0 <=Int WIDTH    [simplification]
-    rule #range((BUF1 +Bytes  BUF2), START, WIDTH) => (#range(BUF1, START                       , lengthBytes(BUF1) -Int START)) +Bytes (#range(BUF2, 0, START +Int WIDTH -Int lengthBytes(BUF1)))  requires                        lengthBytes(BUF1) <Int START +Int WIDTH andBool #range(0 <= START < lengthBytes(BUF1)) [simplification]
+    // Memory update
 
-    rule #range(#padRightToWidth(_, BUF), 0, WIDTH) => BUF requires lengthBytes(BUF) ==Int WIDTH [simplification]
+    rule [memUpdate-is-empty]:
+      _:Bytes [ S:Int := _:Bytes ] => .Bytes
+      requires S <Int 0
+      [simplification]
 
-    // WS [ START := WS' ]
+    rule [memUpdate-as-concat-inside]:
+      B1:Bytes [ S:Int := B2:Bytes ] =>
+        #range(B1, 0, S) +Bytes B2 +Bytes #range (B1 , S +Int lengthBytes(B2), lengthBytes(B1) -Int (S +Int lengthBytes(B2)))
+      requires 0 <=Int S andBool S +Int lengthBytes(B2) <Int lengthBytes(B1)
+      [simplification(60)]
 
-    // re-order memory write
-    rule MEM [ K1 := BUF1:Bytes ] [ K2 := BUF2:Bytes ] => MEM [ K2 := BUF2 ] [ K1 := BUF1 ]
-      requires lengthBytes(BUF2) <=Int K1 -Int K2 andBool K2  <Int K1 [simplification]
+    rule [memUpdate-as-concat-outside-1]:
+      B1:Bytes [ S:Int := B2:Bytes ] => #range(B1, 0, S) +Bytes B2
+      requires 0 <=Int S andBool lengthBytes(B1) <=Int S +Int lengthBytes(B2)
+      [simplification(60)]
 
-    // overwritten memory write
-    rule MEM [ K1 := BUF1 ] [ K2 := BUF2 ] => MEM [ K2 := BUF2 ]
-      requires K1 +Int lengthBytes(BUF1) <=Int K2 +Int lengthBytes(BUF2) andBool K2 <=Int K1 [simplification]
+    rule [memUpdate-reorder]:
+      B:Bytes [ S1:Int := B1:Bytes] [ S2:Int := B2:Bytes ] => B [ S2 := B2 ] [ S1 := B1 ]
+      requires S2 <Int S1 andBool S2 +Int lengthBytes(B2) <=Int S1
+      [simplification]
```

Comment:
Counterexample:

- `B == #buf(2, 1)`
- `S1== 1`
- `B1 == #buf(1, 2)`
- `S2 == -1`
- `B2 == #buf(1, 3)`

Then:

- `B [ S1 := B1 ] [ S2 := B2 ] => (B [ S1 := B1 ]) [ -1 := #buf(1, 3) ] => .Bytes`
- `B [ S1 := B1 ] [ S2 := B2 ] => B [ S2 := B2 ] [ S1 := B1 ] => (B [ -1 := #buf(1, 3) ]) [ S1 := B1 ] => .Bytes [ S1 := B1 ] => #buf(1, 0) +Bytes #buf(2, 1)`

Reply:
I think restricting `0 <=Int S2` should solve it.

Reply:
Any counterexamples produced would be good to put in as new `runLemma => doneLemma` style claims.

Reply:
Yes, the (now) usual `0 <=Int ...` - will be fixed with next push.

Reply:
> Any counterexamples produced would be good to put in as new `runLemma => doneLemma` style claims.

I'm not sure what that would entail here?

Reply:
Something like `runLemma(#buf(2, 1) [ 1 := #buf(1, 2) ] [ -1 := #buf(1, 3) ]) => doneLemma(.Bytes)` perhaps?

Reply:
Done and passing.